presto-main/src/main/java/com/facebook/presto/operator/scalar/FailureFunction.java

```diff
@@ -31,12 +31,35 @@ private FailureFunction() {}
 
     // We shouldn't be using UNKNOWN as an explicit type. This will be fixed when we fix type inference
     @Description("Decodes json to an exception and throws it")
-    @ScalarFunction(hidden = true)
+    @ScalarFunction(value = "fail", hidden = true)
     @SqlType("unknown")
-    public static void fail(@SqlType(StandardTypes.JSON) Slice failureInfoSlice)
+    public static void failWithException(@SqlType(StandardTypes.JSON) Slice failureInfoSlice)
     {
         FailureInfo failureInfo = JSON_CODEC.fromJson(failureInfoSlice.getBytes());
         // wrap the failure in a new exception to append the current stack trace
         throw new PrestoException(StandardErrorCode.GENERIC_USER_ERROR, failureInfo.toException());
     }
+
+    @Description("Throws an exception with a given message")
+    @ScalarFunction(value = "fail", hidden = true)
+    @SqlType("unknown")
+    public static void fail(@SqlType(StandardTypes.VARCHAR) Slice message)
+    {
+        throw new PrestoException(StandardErrorCode.GENERIC_USER_ERROR, message.toStringUtf8());
+    }
+
+    @Description("Throws an exception with a given error code and message")
+    @ScalarFunction(value = "fail", hidden = true)
+    @SqlType("unknown")
+    public static void fail(
+            @SqlType(StandardTypes.INTEGER) long errorCode,
+            @SqlType(StandardTypes.VARCHAR) Slice message)
+    {
+        for (StandardErrorCode standardErrorCode : StandardErrorCode.values()) {
+            if (standardErrorCode.toErrorCode().getCode() == errorCode) {
+                throw new PrestoException(standardErrorCode, message.toStringUtf8());
+            }
+        }
+        throw new PrestoException(StandardErrorCode.GENERIC_USER_ERROR, "Unable to find error for code: " + errorCode);
```

Comment:
Internal presto error (e.g: invalid error code) is mixed with intentional query error. Maybe let's just have a dedicated failure method for subquery? What do you think?

Reply:
I change the error code here to `GENERIC_INTERNAL_ERROR`. I think special method for single use case sounds like an overkill. Idk.